### PR TITLE
dynamic_forward_proxy: fix sub_cluster_confg stuck with warm up due to DFPCluster removed

### PR DESCRIPTION
**Pull request by itsLucario**
Commit Message:
- Add an optional argument called `ignore_removal` to `ClusterManagerImpl::addOrUpdateCluster` and `ClusterManagerImpl::loadCluster`. This argument defaults to `false`, so it won't affect any existing flows. We'll be setting this in Cluster Data. 
- If the `ignore_removal` value is `true`, then the cluster won't be removed when `ClusterManagerImpl::removeCluster` is called.
- To remove a cluster that has been added with `ignore_removal` set to `true`, the `remove_ignored` argument must be set to `true` when `ClusterManagerImpl::removeCluster` is called.

This helps to manage clusters whose lifecycle is managed by custom implementations similar to DFP clusters.

Additional Description:
- Currently clusters that are dynamically added with custom implementations from filters, etc. are removed when a CDS event is triggered. This is because these dynamically created clusters will come in diff of CDS and will be removed by the cluster manager.

Risk Level: low
Testing: unit test
Docs Changes: no
Release Notes: no
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

Fixes: #35171

### CODE REVIEW

**Comment by itsLucario** [comment]
@adisuissa Can you please help with review?

File: api/envoy/config/cluster/v3/cluster.proto
Comment (adisuissa) [inline api/envoy/config/cluster/v3/cluster.proto:1227]: I'm not sure this is the right place to make this change.
Shouldn't this be part of the dynamic-forward-proxy config, and not per-cluster?
The risk is that clusters are will linger for a long time, and it is unclear why would one want to keep them around with no way to remove.

**Review by adisuissa — COMMENTED** [review]
Thanks for contributing.
Left a high-level question.

File: api/envoy/config/cluster/v3/cluster.proto
Comment (itsLucario) [reply api/envoy/config/cluster/v3/cluster.proto:1227]: Thanks. That's a valid question. Yes, this can be part of dynamic-forward-proxy, and we can instruct it to recreate the DFP cluster if it's not found (in case it was removed by the cluster manager).

However, in environments like service-mesh or configurations where CDS can be a very frequent event, the cluster manager will end up removing it very frequently. Also, when used as an egress, DFPClusters can be high in number, and removing and recreating them on CDS after the cluster manager removes them can affect performance.

In use cases like dynamic-forward-proxy, where clusters are created and managed by a custom implementation (create/update/remove), CDS should not impact them. dynamic-forward-proxy has TTL which does remove cluster if not used for the TTL duration. 

We can hide the field from docs and let this to be used for clusters managed by custom implementations.

Open for feedback and thoughts.

**Comment by ravenblackx** [comment]
Ping @adisuissa I think this is waiting for opinion from you.

File: api/envoy/config/cluster/v3/cluster.proto
Comment (doujiang24) [reply api/envoy/config/cluster/v3/cluster.proto:1227]: > hide the field from docs 

Personally, I think it could be a choice, and note clearly in comment: it's just for internal usage, not for users.

Another choice is: we add one more argument, `ignore_removal`, in `ClusterManagerImpl::addOrUpdateCluster` and `ClusterManagerImpl::loadCluster`.
It should works too, but with more code base changes.

Comment (itsLucario) [reply api/envoy/config/cluster/v3/cluster.proto:1228]: Right! @doujiang24 The later one needs more changes. Leaving note and hide from docs is simpler. 
Made changes: Note in the comment and hide it from docs. 

Open for thoughts

Comment (adisuissa) [reply api/envoy/config/cluster/v3/cluster.proto:1228]: Hiding from the docs should not be considered, IMHO, as it should be used for API that needs more testing (not to be hidden from some use-cases).
I think the question is whether it makes sense for DFP to have both clusters that are removed and clusters that are kept all the time. From your comments I gather that in DFP one would want to ignore the CDS removal, so it seems that a single knob (on the DFP config) makes more sense, no?

**Review by adisuissa — COMMENTED** [review]
Apologies, I was under the weather, and just got back to this.

File: api/envoy/config/cluster/v3/cluster.proto
Comment (itsLucario) [reply api/envoy/config/cluster/v3/cluster.proto:1228]: On CDS event when clusters are being removed, the cluster manager cannot differentiate whether it is a DFP cluster or a normal cluster. 

There is nothing to differentiate and prevent these DFP clusters from getting removed by cluster manager.

Please correct me if I'm wrong or not making sense.

Comment (doujiang24) [reply api/envoy/config/cluster/v3/cluster.proto:1228]: The lifecycle of the sub clusters are managed by the DFP cluster, both creation and deletion.
So, we want to add a flag for the sub clusters to not be removed by other events.

The flag do not have to be a configuration, just a internal flag.

If hidding from docs is not be considered, maybe we can try the option 2？ https://github.com/envoyproxy/envoy/pull/35848#discussion_r1759645687 cc @itsLucario

Comment (itsLucario) [reply api/envoy/config/cluster/v3/cluster.proto:1228]: Right! Summarizing the other approach. 

We'll remove the `ignore_removal` attribute addition from Cluster API Spec.
Introduce an optional argument `ignore_removal` default set to `false` (optional param to ensure existing flow won't break) in `ClusterManagerImpl:: addOrUpdateCluster` and `ClusterManagerImpl::loadCluster` so that ClusterData can set this value and use can check while removing.

Does this sound right! If people are fine, I'll proceed with working on changes.

File: source/common/upstream/cluster_manager_impl.cc
Comment (doujiang24) [inline source/common/upstream/cluster_manager_impl.cc:887]:
```suggestion
      （!existing_active_cluster->second->ignore_removal_  || remove_ignored)) {
```

does this could be better?

Comment (itsLucario) [reply source/common/upstream/cluster_manager_impl.cc:887]: This will make cluster to remove when remove_ignored set to true and cluster data has ignore_removal set to false.

This should fine as well. removeCluster will behave as current for cluster with ignore_removal false irrespective of remove_ignored value. And remove_ignored will only take effect for cluster with ignore_removal set to true.

File: api/envoy/config/cluster/v3/cluster.proto
Comment (adisuissa) [reply api/envoy/config/cluster/v3/cluster.proto]: FWIW, I didn't say that removing the API field is required, I just suggested moving it to the DFP config instead of having it in a per-Cluster config.

Taking a step back, there are a few things to clarify:
1. Changing the API should be done in a considerate manner, as it impacts not only Envoy users, but also xDS-server maintainers. Moreover, removing fields essentially cannot be done, so we prefer to get it mostly right now. That is why we are discussing if a field should be added, and if so where.
2. If a PR suggests changing the way Envoy behaves right now (breaking behavior change), this can either be done by adding a field or a runtime-flag. The former is a long-term support, say a filter that needs to support 2 modes of operation. The latter is when a bug is observed, and the fix causes a user-facing change. Runtime-flags allow Envoy users ~6 months to gradually ensure that the other systems behave correctly with the new change.

In the context of this PR, it is unclear to me whether it should be a config-enabled feature or a runtime-flag guarded change. I'm leaning towards the runtime guarded change according to your description (but could be persuaded towards an API feature if there are use-cases where the clusters that are added by DFP need to removed upon a CDS update).

Comment (itsLucario) [reply api/envoy/config/cluster/v3/cluster.proto]: Thanks for the thoughts!
1. Totally agreed.  Chaging API should be done in considerate manner. After your thoughts it makes more sense to follow different approach. I feel the approach mentioned by @doujiang24 is fiesible and easy to use even in future usecases.
2. The PR is not going to change any current behaviour and doesn't break any current behaviour as well. This should be straight enough that we want to avoid removal of cluster by cluster manager whose lifecycle is managed by custom implementation.

I have updated the description accordingly.

File: source/common/upstream/cluster_manager_impl.h
Comment (adisuissa) [inline source/common/upstream/cluster_manager_impl.h:767]: I think that the name `ignore_removal_` is somewhat ambiguous. Specifically, it is unclear what should happen when the `remove_ignored` argument of `removeCluster()` is `true` while `ignore_removal_` is also true (I understand that it is specified in the method's comment, but there may be a way to improve upon this).
Please rename to something like `avoid_cds_removal_`, which specifies that it should only avoid CDS removals.

**Review by adisuissa — COMMENTED** [review]
Overall LGTM, thanks!
Please add a release note.
I'm not sure if this needs a runtime-guard or not (I tend to have one because it is a user-facing change, but according to the Envoy's contributors guide it seems not to be mandatory in this case). I'll let the senior-maintainer decide on that.

Assigning @alyssawilk as DFP code-owner.
/assign @alyssawilk

File: source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc
Comment (alyssawilk) [inline source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc:94]: Makes sense that we don't want DFP sub-clusters removed by a CDS update, but just to check my understanding it is plausible that a non-sub-cluster DFP cluster be removed via CDS right? If so let's make sure we test that, and either way let's runtime guard this change.

File: source/common/upstream/cluster_manager_impl.h
Comment (itsLucario) [reply source/common/upstream/cluster_manager_impl.h:767]: Right! `avoid_cds_removal_` sounds better. 
For `remove_ignored` argument of removeCluster(), how if we rename it `force_remove`?

**Review by adisuissa — DISMISSED** [review]
LGTM, thanks!

**Comment by adisuissa** [comment]
I'm trying to rerun the job, but it may need another main merge (due to a recent CI upgrade)

**Comment by adisuissa** [comment]
Seems that CI [complains](https://github.com/envoyproxy/envoy/actions/runs/11157572663/job/31012169240#step:12:5858) about coverage. Please follow link at https://github.com/envoyproxy/envoy/actions/runs/11157572663/job/31012169240#step:12:6358 to see the code that is not covered.
/wait

**Comment by itsLucario** [comment]
Seems like CI is now failing due to [Invalid Creds](https://github.com/envoyproxy/envoy/actions/runs/11181916177/job/31129666841#step:12:1061)

File: source/extensions/clusters/dynamic_forward_proxy/cluster.cc
Comment (adisuissa) [inline source/extensions/clusters/dynamic_forward_proxy/cluster.cc:100]: nit: `cm_.removeCluster(cluster_name, Runtime::runtimeFeatureEnabled("envoy.reloadable_features.avoid_dfp_cluster_removal_on_cds_update"));`

File: source/extensions/clusters/dynamic_forward_proxy/cluster.cc
Comment (adisuissa) [inline source/extensions/clusters/dynamic_forward_proxy/cluster.cc:144]: Same as above.

File: source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc
Comment (adisuissa) [inline source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc:91]: Same as above.

**Review by adisuissa — COMMENTED** [review]
Thanks!
LGTM, and left some minor nits.
Good to merge IMHO.

File: source/common/upstream/cluster_manager_impl.cc
Comment (alyssawilk) [inline source/common/upstream/cluster_manager_impl.cc:886]: Thanks for picking up this change, and sorry coming in late with a possibly naive question.
So the problem is we add a DFP cluster via CDS, the DFP cluster adds each host as a sub-cluster, and subsequently the DFP sub-clusters are removed by CDS updates right?

Would it be cleaner rather than adding the ignore tags, to simply have the sub-clusters not tagged as added via API?  Or would that then cause problems for removal?

If it complicates removal WDYT of instead of having an "added by API" bool and a "ignore" bool, if we change both addOrUpdateCluster to include a enum Type (Static, Cds, DFP) where each cluster retains its type and we ignore CDS removals for DFP?

/wait-any for comments

Comment (itsLucario) [reply source/common/upstream/cluster_manager_impl.cc:886]: Thanks for your inputs. 

Tagging clusters with `not added_via_api` would make the changes more complex and could impact existing behavior. 

The cluster api spec already has `type` and `cluster_type` fields, but the DFP sub-cluster sets `type` to `STRICT_DNS`. Having another enum in API spec or having enum as addOrUpdateCluster param will abstract the resulting behavior, i.e., avoiding cds removal.

I feel `avoid_cds_removal` parameter makes it clear & clean for the caller and easy to manage, not just for DFP sub-clusters, but for any other use cases in future where custom implementations will manage the cluster lifecycle (similar to DFP).

Open for thoughts :)

**Review by alyssawilk — APPROVED** [review]
fair enough!

**Pull request merged**
